### PR TITLE
PLANNER-2315 Properly size Bavet child tuple lists

### PR DESCRIPTION
**Pull request by triceo**
Some tuples will only ever have 1 child.
Some tuples will mostly only have 1 expected child.
Some tuples will never have children.

The sizing of the collections should reflect that, which leads to a performance improvement of ~10 % on average.

### CODE REVIEW

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/bi/BavetGroupBiTuple.java [hunk 1]
```diff
@@ -29,6 +32,7 @@
     private int parentCount;
     private ResultContainer_ resultContainer;
     private Result_ result;
+    private final List<BavetAbstractTuple> childTupleList = new ArrayList<>();
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/bi/BavetGroupBiTuple.java:35]: The bridge will have n childs typically, but the group/joins itself - like this one - will have 1 child typically.

```suggestion
    private final List<BavetAbstractTuple> childTupleList = new ArrayList<>(1);
```

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/uni/BavetFromUniTuple.java [hunk 2]
```diff
@@ -16,12 +16,19 @@
...
+import org.optaplanner.core.impl.score.stream.bavet.common.BavetAbstractTuple;
+
 public final class BavetFromUniTuple<A> extends BavetAbstractUniTuple<A> {
 
     private final BavetFromUniNode<A> node;
     private final A factA;
+    private final List<BavetAbstractTuple> childTupleList;
 
     public BavetFromUniTuple(BavetFromUniNode<A> node, A factA, int childTupleListSize) {
+        this.childTupleList = new ArrayList<>(childTupleListSize);
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/uni/BavetFromUniTuple.java:31]: Idea: as we apparrantly know the maximum childTupleListSize for a fromUni tuple,
we probably know it for the filterUni tuples too?

For example:
- from(A)
-- filter(A.b != null)
---  join(C).penalize
--- join(D).penalize
--- filter(E).penalize
-- filter(F).penalize

the from has max 2 child tuples, the filter has max 3. Even if the turns out less that than, the 3 is still a very good ArrayList capacity.

**Review by ge0ffrey — APPROVED** [review]

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/uni/BavetFromUniTuple.java (hunk 2 above)
Comment (triceo) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/uni/BavetFromUniTuple.java:31]: You keep assuming that there is node sharing. That is a very poor assumption, IMO, outside of the from nodes.
The bar to actually get a node shared is very high, and people will never meet it.

(Unless they write some complicated and non-straight-forward constraint providers.)

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/bi/BavetGroupBiTuple.java (hunk 1 above)
Comment (triceo) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/bi/BavetGroupBiTuple.java:35]: Hmm, I have it switched the other way. Will need rebenchmarking then, could be even better.

**Pull request merged**